storage/merge.go

```diff
@@ -153,13 +153,21 @@ func (q *mergeGenericQuerier) Select(ctx context.Context, sortSeries bool, hints
 	)
 	// Schedule all Selects for all queriers we know about.
 	for _, querier := range q.queriers {
+		// copy the matchers as some queriers may alter the slice.
```

Comment:
I don't think that preventing blockQuerier from altering the matchers or assuming no other querier will respect that in the future is a good/safe solution.